### PR TITLE
Fix/rz find r option double free

### LINKED ISSUE #5189: Allow `rz-find` to execute also system shell command on match
Currently we have `rz-find -E` but it runs Rizin's shell commands. To become more useful for scripting scenarios, `rz-find` should also have the command to run system shell commands (Bash, ZSH, PowerShell, etc).

Maybe `rz-find -R` as in "R"un?

Another option would be swap them:
- `rz-find -E ...` would **E**xecute shell command
- `rz-find -R ...` would run **Rizin** command

It also should allow passing the matching file name to that command, obviously.

See files:
- `binrz/rz-find/*`
- `librz/main/rz-find.c`

### PR DESCRIPTION
**Pull request by Matulpatel**
**Your checklist for this pull request**
-  I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
-  I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).
- [ ] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**
This PR introduces the -R option in rz-find, enabling execution of arbitrary system shell commands on each matched file, with {} dynamically replaced by the filename. Currently, rz-find supports the -E option to execute Rizin shell commands, but scripting scenarios often require running native system shell commands.
The newly added -R option fills this need, improving usability by allowing external shell commands (Bash,PowerShell, etc.) to be run, with the matched filename passed to the command.

Implementation Details

Parses and stores the -R command string safely by duplicating it. Inside the hit() callback, duplicates the command string before replacement, and uses rz_str_replace() to swap {} with the current filename, and executes the command via rz_sys_system().
Tried to ensure to avoids double frees and crashes.

Testing
Verified command substitution and execution on directory trees with multiple files.
Confirmed no crashes and correct system command output.
Used AddressSanitizer  to ensure memory safety.
Tested edge cases: empty commands, no matches, special characters.

**Test plan**

To verify:
1. Run: ./build/binrz/rz-find/rz-find -s "teststring" -R "echo filename: {}" test_rzfind/
   - Should display filename for each match.
2. Try with files containing multiple matches and special character names.
3. AddressSanitizer confirms no memory leaks.
Screenshot attached below show multi-file, multi-match scenario and output:

Screenshot ...

**Closing issues**
closes #5189

...
Screenshot 2025-11-02 at 5 08 36 PM

### CODE REVIEW

File: librz/main/rz-find.c [hunk 1]
```diff
@@ -160,6 +161,30 @@ static int hit(RzSearchKeyword *kw, void *user, ut64 addr) {
 		free(command);
 		return 1;
 	}
+	if (ro->sys_command) {
+		// Duplicate the sys_command string to pass ownership to rz_str_replace
```
Comment (Rot127) [inline librz/main/rz-find.c]:
```suggestion
```

File: librz/main/rz-find.c [hunk 2]
```diff
@@ -160,6 +161,30 @@ static int hit(RzSearchKeyword *kw, void *user, ut64 addr) {
...
+		char *tmp = rz_str_dup(ro->sys_command);
+		if (!tmp) {
+			// Handle allocation failure
```
Comment (Rot127) [inline librz/main/rz-find.c]:
```suggestion
```

File: librz/main/rz-find.c [hunk 3]
```diff
@@ -160,6 +161,30 @@ static int hit(RzSearchKeyword *kw, void *user, ut64 addr) {
...
 		return 1;
 	}
+	if (ro->sys_command) {
+		// Duplicate the sys_command string to pass ownership to rz_str_replace
+		char *tmp = rz_str_dup(ro->sys_command);
+		if (!tmp) {
+			// Handle allocation failure 
+			return 1;
+		}
+		// Now safely call rz_str_replace which will free tmp internally
```
Comment (Rot127) [inline librz/main/rz-find.c]:
```suggestion
```

File: librz/main/rz-find.c [hunk 4]
```diff
@@ -160,6 +161,30 @@ static int hit(RzSearchKeyword *kw, void *user, ut64 addr) {
...
+				return 1;
+			}
+		}
+		int status = rz_sys_system(replaced);
+		if (status == -1) {
+			RZ_LOG_ERROR("Failed to execute command: %s\n", replaced);
+		}
+		if (replaced) {
+			free(replaced);
+		}
```
Comment (Rot127) [inline librz/main/rz-find.c]:
```suggestion
		free(replaced);
```

File: librz/main/rz-find.c [hunk 5]
```diff
@@ -656,5 +686,9 @@ RZ_API int rz_main_rz_find(int argc, const char **argv) {
 	if (ro.json) {
 		printf("]\n");
 	}
+	if (ro.sys_command) {
+    	free((void *)ro.sys_command);
+    	ro.sys_command = NULL;
```
Comment (Rot127) [inline librz/main/rz-find.c:692]:
```suggestion
    	RZ_FREE(ro.sys_command);
```

File: test_rzfind/file1.txt
Comment (Rot127) [inline test_rzfind/file1.txt:1]: Remove leftovers

**Review by Rot127 — CHANGES_REQUESTED** [review]
Please add a test in `test/db/tools/rz_find`, it doesn't matter if the actual system command fails, we just need to cover the paths.

**Comment by Rot127** [comment]
Also set a proper title for the PR please

**Review by notxvilka — CHANGES_REQUESTED** [review]
As `-E` was introduced relatively recently and likely nobody used it yet, I suggest swapping `-E` with `-R`

**Comment by Matulpatel** [comment]
Sure will do that

File: librz/main/rz-find.c [hunk 6]
```diff
@@ -200,7 +226,8 @@ static int show_help(const char *argv0, int line) {
 		"-a",    "align",   "Only accept aligned hits",
 		"-b",    "size",    "Set block size",
 		"-e",    "regex",   "Search for regex matches (can be used multiple times)",
-		"-E",    "cmd",     "Execute command for each file found",
+		"-R",    "cmd",     "Execute command for each file found",
```
Comment (notxvilka) [inline librz/main/rz-find.c:229]: "Execute rizin command..."

**Review by notxvilka — CHANGES_REQUESTED** [review]
Test changes make no sense, only `rz-find` one should be changed.
Also, please add a test for `rz-find -E` with something like `whoami` or `date` that are present on each system but aren't part of the Rizin shell syntax.

File: test/db/tools/rz [hunk 7]
```diff
@@ -181,31 +181,486 @@ NAME=b/w prompt
...

[0x00000000]> 
[0x00000000]> [2K
[2K

[0x00000000]> q
[0x00000000]> q[2K

[0x00000000]> q
+[2K
```
Comment (Rot127) [inline test/db/tools/rz:184]: Why the new lines?

**Review by Rot127 — CHANGES_REQUESTED** [review]

File: test/db/tools/rz_find [hunk 8]
```diff
@@ -263,3 +263,12 @@ EXPECT=<<EOF
 1
 EOF
 RUN
+
+NAME=rz-find -E runs system command
```
Comment (notxvilka) [inline test/db/tools/rz_find:267]: Wrong test, see how previous is checked. Also, previous one should be fixed by using `-R` instead of `-E`

File: test/db/tools/rz
Comment (notxvilka) [inline test/db/tools/rz:1]: Undo changes in this file completely

**Review by notxvilka — CHANGES_REQUESTED** [review]

File: test/db/tools/rz_find (hunk 8 above)
Comment (Matulpatel) [reply test/db/tools/rz_find:267]: oh ok thanks , will resolve it

**Comment by notxvilka** [comment]
Closing due to inactivity
